src/bookshelf/src/bksw_sec.erl

```diff
@@ -21,123 +22,405 @@
+
+        % CODE REVIEW: used in both verification types
+        RawMethod = wrq:method(Req0),
+        Method = list_to_atom(string:to_lower(erlang:atom_to_list(RawMethod))),
+
+        % CODE REVIEW: used in both verification types
+        Path = wrq:path(Req0),
+
+        % CODE REVIEW: used in both verification types
+        Config = mini_s3:new(AccessKey, SecretKey, Host),
+
+        % CODE REVIEW: AltSignedHeaders used in both verification types
+
+        % replace host header with alternate host header
+        AltHost = mini_s3:get_host_toggleport(Host, Config),
+        AltSignedHeaders = [case {K, V} of {"host", _} -> {"host", AltHost}; _ -> {K, V} end || {K, V} <- SignedHeaders],
+
+        % CODE REVIEW: used in both verification types
+        XAmzExpires = list_to_integer(XAmzExpiresString),
+        case XAmzExpires > 1 andalso XAmzExpires < 604800 of
+            true -> ok;
+            _    -> throw({RequestId, Req0, Context})
+        end,
+
+        CalculatedSig = case VerificationType of
+            presigned_url ->
+
+                case wrq:get_qs_value("X-Amz-Algorithm", Req0) of
+                    "AWS4-HMAC-SHA256" -> ok;
+                    _                  -> throw({RequestId, Req0, Context})
+                end,
+
+                case check_signed_headers_common(SignedHeaders, Headers) of
+                    true -> ok;
+                    _    -> throw({RequestId, Req0, Context})
+                end,
+
+                {Bucketname, Key} = get_bucket_key(Path),
+
+                ComparisonURL = mini_s3:s3_url(Method, Bucketname, Key, XAmzExpires, SignedHeaders, Date, Config),
+
+                % list_to_binary profiled faster than binary_to_list,
+                % so use that for conversion and comparison.
+                IncomingSig = list_to_binary(IncomingSignature),
+
+                [_, ComparisonSig] = string:split(ComparisonURL, "&X-Amz-Signature=", trailing),
+
+                % TODO: try to remove alt sig computation and see what happens
+                % NOTE: this was tried, and caused compilation and test failures.
+
+                case IncomingSig of
+                    ComparisonSig ->
+                        %AltComparisonSig = "not computed",
+                        IncomingSig;
+                    _ ->
+                        AltComparisonURL = mini_s3:s3_url(Method, Bucketname, Key, XAmzExpires, AltSignedHeaders, Date, Config),
+                        [_, AltComparisonSig] = string:split(AltComparisonURL, "&X-Amz-Signature=", all),
+                        AltComparisonSig
+                end;
+            authorization_header ->
+
+                ComparisonURL = "not-applicable",
+                QueryParams = wrq:req_qs(Req0),
+
+                case check_signed_headers_authhead(SignedHeaders, Headers) of
+                    true -> ok;
+                    _    -> throw({RequestId, Req0, Context})
+                end,
```

Comment:
Can be done as a part of malformed_request

Reply:
Done.